enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,198 @@
+
+PROs:
+
+- we own it, can add anything that is needed
+
+CONs:
+
+- still vaporware
+
+A mutating webhook (plus a  controller) will inspect created resources in specified namespaces for labeling scheme compliance. The webhook will check the namespace, type, ... of the created resource, match it to the configured labeling scheme, and add any missing labels/annotations. The webhook could also update incorrectly set labels/annotations if requested.
+
+The webhook will be configured by means of a LabelSchemeConfig CRD. The LabelSchemeConfig CRD will specify a target namespace, allowing to specify a label scheme for the specific namespace (an alternative solution of having the LabelSchemeConfig CRD namespaced would allow project admins to access the resource, which is not desirable).  
+ The controller will look for LabelSchemeConfig CRDs in it's own namespace.
+
+A sample LabelSchemeConfig resource is shown below:
+
+```yaml
+apiVersion: v1
+kind: LabelSchemeConfig
+metadata:
+ name: labeling-scheme-config-for-namespace1
+ namespace: controller-namespace
+spec:
+ labelSchemeNamespace: namespace1
+ labelScheme:
+   - apiVersion: networking.k8s.io/v1/
+     kind: NetworkPolicy
+     labelSelector: ""
+     labels:
+       label1-key: label1-value
+       label2-key: label2-value
+     annotations:
+       annotations1-key: annotations1-value
+   - apiVersion: v1
+     kind: Pod
+     labelSelector:
+       - app: custom-app
+     resourceName: resource-name
+     labels:
+       - label1-key: label1-value
+       - label2-key: label2-value
+     annotations:
+       annotations1-key: annotations1-value
+```
+
+The LabelSchemeConfig CRD will require a separate resource for each namespace, which although elegant migth be laborsome to maintain. An alternative solution could be to allow to specify a list of LabelSchemeConfig's in one resource:
+
+```yaml
+apiVersion: v1
+kind: LabelSchemeConfig
+metadata:
+ name: labeling-scheme-config-for-multiple-namespaces
+ namespace: controller-namespace
+spec:
+  - labelSchemeNamespace: namespace1
+    labelScheme:
+      - apiVersion: networking.k8s.io/v1/
+        kind: NetworkPolicy       
+  - labelSchemeNamespace: namespace2
+    labelScheme:
+  ...
+```
+
+#### Resource relabeling
+
+The labeling scheme could change during the lifecycle of labeled resources. The resources in question could either be left as is, retaining the old labels, or relabeled.
+
+Relabeling could be an issue for resources such as pods, if the podinfo has been used for pod configuration. For telco applications this could present an issue, if for example a sidecar used for the service mesh is making use of the labeling.
+
+A relabeling policy will be added to the configuration to allow the user to decide on the configuration:
+
+```yaml
+ ...
+ labelScheme:
+   ...
+   - apiVersion: v1
+     kind: Pod
+     relabelPolicy: No | Yes 
```

Comment:
how can a policy be no or yes? `policy: Relabel` or `relabel: yes` would make sense.

Reply:
Changed to "relabel".